[MID-3] add header footer components

Yivgha opened the pull request:
+tailwind styles;
+reusable btn component;

trello: https://trello.com/c/Ll5qjjEM

env: https://mayak-2-h8fflutym-yivghas-projects.vercel.app/

Yivgha committed:
add header&footer components

+tailwind styles;
+reusable btn component;

Yivgha committed:
run linter fix

tried to fix some errors with quotes

Dimirolz commented on src/app/_components/Footer/Footer.js:
```diff
@@ -1,5 +1,101 @@
+
+  return (
+    <footer className="text-white flex w-full flex-col bg-primary-800 px-20 py-12">
+      <div className={`${flexBetween}`}>
```
@Yivgha let's refactor all the class interpolation into the usage of the `cn` utility.

Take inspiration from here -> https://github.com/shadcn-ui/ui/blob/0fae3fd93ae749aca708bdfbbbeddc5d576bfb2e/templates/next-template/lib/utils.ts#L2

TwMerge is an important piece of the workflow with Tailwind. It fixes all the problems related to Tailwind style conflicts.

Yivgha replied:
Good advice! Created cn util fn.

Dimirolz commented on src/app/_components/Footer/Footer.js:
```diff
@@ -1,5 +1,101 @@
+            Слідкуй за нами в соцмережах
+          </p>
+
+          <Link
```
@Yivgha let's extract this link repetition into a loop. You can create an array, socialLinks, with objects that contain information for each link.

The `mr-4` between links is doing the job, but it's more convenient to use `gap-4` and put the links inside a flex container.

Yivgha replied:
created reusable link components

Dimirolz commented on src/app/_components/Header/Header.js:
```diff
@@ -1,5 +1,110 @@
+            </Link>
+          </li>
+        </ul>
+        <div className={`${flexBetween} gap-6`}>
```
@Yivgha I saw similar links in the footer, so I suggest you extract this markup with social links into a separate component, like `SocialLinks`.

Yivgha replied:
done

Dimirolz commented on src/app/_components/Icons/Facebook.js:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import p from 'prop-types';
+
+const FacebookComponent = ({ className }) => (
```
@Yivgha for such cases, you can often see the usage of `svgr `in various projects. I don't have a strong opinion there; `svgr `can be tricky to set up sometimes.

You can discuss the need for `svgr ` in such cases with your colleagues.

Yivgha replied:
implemented svgr

Dimirolz commented on src/app/_components/OutlinedBtn/OutlinedBtn.js:
```diff
@@ -0,0 +1,25 @@
+      type="button"
+      className={`${className} ${flexCenter} ${basicBtn} ${btnHoverFocus} ${transition}`}
+    >
+      {' '}
```
@Yivgha I'm not sure what the reason is for those white spaces. It would be better to remove them from the base component.

Yivgha replied:
It was linter fix, deleted.

Dimirolz reviewed (changes requested):
Please resolve the conflicts of this branch with the current `dev` branch

alex-popov-tech commented on src/app/_components/Footer/Footer.js:
```diff
@@ -1,5 +1,101 @@
-import React from 'react';
+"use client";
```
🔴 why is it client component? i don't see any client specific features used here

Yivgha replied:
Deleted.

alex-popov-tech commented on src/app/_components/Footer/Footer.js:
```diff
@@ -1,5 +1,101 @@
+import React from "react";
+import Image from "next/image";
+import Link from "next/link";
+import OutlinedBtn from "../OutlinedBtn";
+import * as FooterImg from "../../../../public/assets/icons/white-logo.svg";
+import InstagramComponent from "../Icons/Instagram";
+import FacebookComponent from "../Icons/Facebook";
+import XComponent from "../Icons/XSocial";
```
🟠 can these imports be prettified with import aliases?

Yivgha replied:
added some aliases

alex-popov-tech commented on src/app/_components/Header/Header.js:
```diff
@@ -1,5 +1,110 @@
+'use client';
```
🟠 comments about footer are relevant for header component too

Yivgha replied:
Deleted.

alex-popov-tech commented on src/app/_components/Icons/Facebook.js:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import p from 'prop-types';
+
+const FacebookComponent = ({ className }) => (
```
🟢 can't we use something like react-icons for this?

Yivgha replied:
I'd like to, but react-icons don't have all icons we need.

alex-popov-tech commented on src/app/_components/OutlinedBtn/index.js:
```diff
@@ -0,0 +1 @@
+export { default } from './OutlinedBtn';
```
🟠 what is the point for component of 20 lines long to create folder and 2 files?

Yivgha replied:
Deleted.

alex-popov-tech commented on src/app/admin/adminApp.js:
```diff
@@ -38,13 +38,13 @@ export default function AdminPage() {
       />
       <Resource
-        name='Specialist'
+        name="Specialist"
```
🟠 is that according to linter rools? next file uses single quotes, i suspect there is something going on :)

Yivgha replied:
added some .prettierrc.json changes

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into origin/MID-3_add-header-footer

Yivgha committed:
add styles and component refactoring

+ fix linter quotes & CRLF problem;
+ add configs in prettierrc.json;
+ add "tailwind-merge";
+ add cn lib & fn in utils;
+ add  import aliases;
+ add socialLink template;

Yivgha committed:
refactor social icons

+create custom reusable link for socials

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into origin/MID-3_add-header-footer

Yivgha committed:
add reusable links to header&footer

Yivgha committed:
path changes

+lil styles fix

Dimirolz commented on jsconfig.json:
```diff
@@ -1,7 +1,13 @@
+      "@utils/*": ["./src/app/utils/*"],
+      "@config/*": ["./src/app/config/*"],
+      "@socialLink/*": ["./src/app/_components/SocialLink/*"],
+      "@innerLink/*": ["./src/app/_components/InnerLink/*"]
```
@Yivgha you can create links only to _components, not for each individual component, and have paths like `@components/SocialLink`. It's more convenient than creating them for each component.

Dimirolz commented on src/app/_components/SocialLink/SocialLink.js:
```diff
@@ -0,0 +1,61 @@
+
+export default function SocialLink({ items, className, status }) {
+  const { links } = items;
+  return status === 'footerSocials' ? (
+    <div className="flex gap-4">
+      {links
+        ?.sort((a, b) => {
+          if (a.title < b.title) {
+            return -1;
+          }
+          if (a.title > b.title) {
+            return 1;
+          }
+          return 0;
+        })
+        .map((link, idx) => (
+          <Link
```
@Yivgha Let's refactor this component in a few steps.

1. First, let's make it dumb. Simply render links without any sorting or branching logic. Move the sorting logic into the `Footer` component.
2. Remove `status` prop
3. Receive only the necessary props. In this case, accept only the links objects. Not whole nav object.
4. Replace the usage of `.sort` with `.toSorted` (or clone array before sorting, like [...arr].sort) since it doesn't create side effects by mutating the original array.

Yivgha replied:
Done.

alex-popov-tech commented on next.config.mjs:
```diff
@@ -11,6 +11,14 @@ const nextConfig = {
     ignoreDuringBuilds: true,
   },
+  webpack(config) {
+    config.module.rules.push({
```
🟢 would be really nice to have comments describing purpose of such changes

alex-popov-tech commented on src/app/_components/Footer/Footer.js:
```diff
@@ -1,9 +1,58 @@
+  const iconColors = 'text-other-white hover:text-primary-400';
+  const transition = 'transition duration-200 ease-in-out';
+  const footerBtnHover = 'hover:bg-gray-100 hover:text-primary-500  ';
+  const footerBtnFocus = 'focus:bg-gray-100 focus:text-primary-600 ';
```
🟢 there are trailing spaces in strings

Yivgha committed:
change path and reusable links

Yivgha committed:
small fix

BigTako commented on src/app/_components/OutlinedBtn/OutlinedBtn.js:
🟠 - can we join SocialLink.js and InnerLink.js into single component?

BigTako replied:
Discussed in private network

BigTako commented on src/app/_components/Header.js:
🟠 - when we use component without children, we can use just `<Component />`, use <InnerLink ... /> and <SocialLink ... /> please

BigTako replied:
Discussed in private conversation.

Yivgha committed:
Merge branch 'dev' of https://github.com/keenethics/mayak into origin/MID-3_add-header-footer

Yivgha committed:
resolve conflicts

alex-popov-tech commented on src/app/_components/OutlinedBtn/OutlinedBtn.js:
```diff
@@ -0,0 +1,39 @@
+import p from 'prop-types';
+import cn from '@utils/cn';
+
+export default function OutlinedBtn({ children, className }) {
```
🟠 isn't button should be part of separate PR with buttons?

alex-popov-tech commented on src/app/_components/Footer.js:
```diff
@@ -0,0 +1,67 @@
+import SocialLink from '@components/Links/SocialLink';
+
+export default function Footer() {
+  const { links } = siteNav;
```
🟠 
1. do you need those links somewhere in another place? if no, can these be placed somewhere near component?
2. please do not mutate data which you store somewhere, its not obvious behaviour for users which might want to update links list

Yivgha committed:
update next.config
